### ISSUE TITLE
Too many open files (CLOSE_WAIT)

### ISSUE THREAD

**Issue by gabrielima** [lostisland/faraday#1279]
I have an app that relies heavily on consuming a third party API. Basically on every user request, every page visited, I make an API request. It's a rather simple app but I have been struggling with a problem that leads to it becoming unresponsive.

I use Typhoeus ut have tried :httpclient too.

This is an example of what I am doing:

```ruby
connection = Faraday.new(url: ENV['API_URL']) do |faraday|
  faraday.adapter :typhoeus, forbid_reuse: true
  faraday.options.timeout = 20
  faraday.options.open_timeout = 10
end

begin
  response = connection.send...
  ...
rescue ...
  ...
ensure
  connection.close
end
```

And on the header of every request I add " Connection: 'close' "

Apparently I have a problem with closing the TCP connection.
When I run the command `lsof | grep CLOSE_WAIT | wc -l` I notice that thousands of files are open in the state `CLOSE_WAIT`

I have changed the number of open files accepted, but that doesn't seem to be the solution for this. It's like a memory leak, I have to close those connections to make my app scalable.
We can have up to 7000 online visitors at a time.

Does anyone know what I could be doing differently ?

**Comment by mattiagiuffrida-st** [comment]
What is that `forbid_reuse: true` you pass to typhoeus?

**Comment by iMacTia** [comment]
Totally! It this was a Rails application, I'd instantiate the Faraday connection in an initializer of some sort, or have a class lazy-creating it the first time and then reusing it.
This will allow you to reuse the adapter connection without the need to close it (assuming your adapter is capable of a persistent connection, but I'd assume Typhoeus is)

Could you give this a try and let me know if you still have the open files issue?

What is that forbid_reuse: true you pass to typhoeus?

**Comment by gabrielima** [comment]
I deployed a patch reusing the connection. It's working, but I am still having a problem with too many open files. Looking deeper into the problem I found that the connections stuck in CLOSE_WAIT are not the ones I make with Faraday (the IP is different). Now I wonder if I had two problems and just fixed one, or if it was something else all along and I didn't have to change the Faraday connection.

Anyways, thank you very much for your help.
